pkg/check/node_cbt.go

```diff
@@ -0,0 +1,100 @@
+package check
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/vmware/govmomi/vim25/mo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/klog/v2"
+)
+
+// CollectNodeCBT emits metric with CBT Config of each VM
+type CollectNodeCBT struct {
+	lastMetricEmission map[string]int
+}
+
+var _ NodeCheck = &CollectNodeCBT{}
+
+const (
+	cbtMismatchLabel = "cbt"
+	CbtProperty      = "ctkEnabled"
+	cbtEnabled       = "enabled"
+	cbtDisabled      = "disabled"
+)
+
+var (
+	cbtMismatchMetric = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Name:           "vsphere_vm_cbt_checks",
+			Help:           "Boolean metric based on whether ctkEnabled is consistent or not across all nodes in the cluster.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{cbtMismatchLabel},
+	)
+)
+
+func init() {
+	legacyregistry.MustRegister(cbtMismatchMetric)
+}
+
+func (c *CollectNodeCBT) Name() string {
+	return "CollectNodeCBT"
+}
+
+func (c *CollectNodeCBT) StartCheck() error {
+	return nil
+}
+
+func (c *CollectNodeCBT) CheckNode(ctx *CheckContext, node *v1.Node, vm *mo.VirtualMachine) error {
+	klog.V(4).Infof("Checking CBT Property")
+
+	propFound := false
+	for _, config := range vm.Config.ExtraConfig {
+		if config.GetOptionValue().Key == CbtProperty {
+			klog.V(2).Infof("Found ctkEnabled property for node %v with value %v", node.Name, config.GetOptionValue().Value)
+			if strings.EqualFold(fmt.Sprintf("%v", config.GetOptionValue().Value), "TRUE") {
+				ctx.ClusterInfo.SetCbtData(cbtEnabled)
+			} else {
+				ctx.ClusterInfo.SetCbtData(cbtDisabled)
+			}
+			propFound = true
+			break
+		}
+	}
+	if !propFound {
+		klog.V(2).Infof("Property not found for node %v", node.Name)
+		ctx.ClusterInfo.SetCbtData(cbtDisabled)
+	}
+
+	return nil
+}
+
+func (c *CollectNodeCBT) FinishCheck(ctx *CheckContext) {
+	cbtData := ctx.ClusterInfo.GetCbtData()
+
+	for k := range c.lastMetricEmission {
+		c.lastMetricEmission[k] = 0
+	}
+
+	klog.V(2).Infof("Enabled (%v) Disabled (%v)", cbtData[cbtEnabled], cbtData[cbtDisabled])
+	if cbtData[cbtEnabled] > 0 && cbtData[cbtDisabled] > 0 {
+		cbtMismatchMetric.WithLabelValues("mismatch").Set(1)
+	} else {
+		cbtMismatchMetric.WithLabelValues("mismatch").Set(0)
+	}
+
+	// Set the counts of enabled vs disabled
+	for cbtEnabled, count := range cbtData {
+		klog.V(4).Infof("CBT (%v): %v", cbtEnabled, count)
+		cbtMismatchMetric.WithLabelValues(cbtEnabled).Set(float64(count))
+	}
```

Comment:
okay, I see that we are trying to emit metric for number of clusters for which metric is enabled and disabled. In which case tbh, we don't need `mismatch` metric, because it can be calculated from these 2 other metric and an alert can be emitted.

Reply:
Do you have an example of how to link the 2 metrics together?  I tried and I am not good at prometheus metrics.  I just figured the counts would also be helpful in them tracking / seeing how many do not match in addition to the logging.

Reply:
It should be possible by simple "AND" of two different expressions in prometheus. We shouldn't have to emit a separate metric for it.

Reply:
Ok, i'll try to make that work.  I tried "AND" and I was getting no results.  I was trying some along the lines of "enabled > 0 and disabled > 0".

Reply:
Ok, I think I got it working by doing: 
`min_over_time(vsphere_vm_cbt_checks{cbt=~"enabled"}[5m]) > 0 and ignoring (cbt) min_over_time(vsphere_vm_cbt_checks{cbt=~"disabled"}[5m]) > 0` .  I'll run some more tests and will make appropriate changes if they all pass.